### PR TITLE
Implement caching layer for data source information

### LINKED ISSUE #916: Implement caching layer for data source information
The data source table should contain
- ID (pk)
- name
- description
- owner
- executable
- request fee
- last update

Also add data source revision table that should contain
- ID (pk)
- Revision number (pk)
- Name
- timestamp
- block height
- tx hash

### PR DESCRIPTION
**Pull request by taobun**
fixed #916

### CODE REVIEW

File: chain/app/db_app.go [hunk 1]
```diff
@@ -25,6 +29,7 @@ func NewDBBandApp(
 ) *dbBandApp {
 	app := NewBandApp(logger, db, traceStore, loadLatest, invCheckPeriod, baseAppOptions...)
 
+	dbBand.ZoracleKeeper = app.ZoracleKeeper
```
Comment (sorawit) [inline chain/app/db_app.go:32]: Oh god.... Do we really need to do this for all keepers?

File: chain/db/data_source.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+		return errors.New("handleMsgCreateDataSource: cannot find data source id")
+	}
+	id, err := strconv.ParseInt(rawID, 10, 64)
+	if err != nil {
+		return err
+	}
+	return b.AddDataSource(
+		id, msg.Name, msg.Description, msg.Owner, msg.Fee, msg.Executable,
+		b.ctx.BlockTime(), b.ctx.BlockHeight(), txHash,
+	)
```
Comment (sorawit) [inline chain/db/data_source.go:80]: I think we also add to revision here.

File: chain/db/db.go [hunk 3]
```diff
@@ -33,6 +42,13 @@ func NewDB(dialect, path string, metadata map[string]string) (*BandDB, error) {
 		"RESTRICT",
 	)
 
+	db.Model(&DataSourceRevision{}).AddForeignKey(
+		"data_source_id",
+		"data_sources(id)",
+		"CASCADE",
```
Comment (sorawit) [inline chain/db/db.go]: Why don't we do RESTRICT? I think it's safer that way.

File: chain/db/data_source.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+		msg.Owner, msg.Fee, msg.Executable, b.ctx.BlockTime(),
+	)
+
+	err := b.tx.Save(&dataSource).Error
+	if err != nil {
+		return err
+	}
+
+	return b.tx.Create(&DataSourceRevision{
+		DataSourceID: int64(msg.DataSourceID),
```
Comment (sorawit) [inline chain/db/data_source.go:99]: Why do we ignore `RevisionNumber`?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/data_source.go (hunk 2 above)
Comment (taobun) [reply chain/db/data_source.go:80]: AddDataSource is a function to create new data source row with revision already.

File: chain/db/data_source.go (hunk 4 above)
Comment (taobun) [reply chain/db/data_source.go:99]: I use auto increment for this field

File: chain/db/db.go (hunk 3 above)
Comment (taobun) [reply chain/db/db.go]: In my opinion, this table will never be deleted by logic, but it's okay to use RESTRICT

File: chain/app/db_app.go (hunk 1 above)
Comment (taobun) [reply chain/app/db_app.go:32]: :(

Comment (taobun) [reply chain/app/db_app.go:32]: If we don't work like this how we send keeper to handleMsg

File: chain/app/db_app.go [hunk 5]
```diff
@@ -101,6 +138,7 @@ func (app *dbBandApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseB
 	res = app.bandApp.BeginBlock(req)
 	// Begin transaction
 	app.dbBand.BeginTransaction()
+	app.dbBand.SetContext(app.DeliverContext)
```
Comment (sorawit) [inline chain/app/db_app.go:141]: Let's also clear this after EndBlock. Also should we make it readonly. Like creating a cache on top and discard after.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/table.go [hunk 6]
```diff
@@ -27,3 +27,22 @@ type ValidatorVote struct {
 	BlockHeight      int64  `gorm:"primary_key;auto_increment:false"`
 	Voted            bool
 }
+
+type DataSource struct {
+	ID          int64 `gorm:"primary_key;auto_increment:false"`
+	Name        string
```
Comment (sorawit) [inline chain/db/table.go]: Let's decorate everything that's not nullable `not null`. We will have better life with GraphQL that way, right @evilpeach ?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/table.go (hunk 6 above)
Comment (evilpeach) [reply chain/db/table.go]: Absolutely

Comment (taobun) [reply chain/db/table.go]: Already did

**Review by sorawit — DISMISSED** [review]

**Review by sorawit — APPROVED** [review]

**Review by evilpeach — APPROVED** [review]

**Pull request merged**